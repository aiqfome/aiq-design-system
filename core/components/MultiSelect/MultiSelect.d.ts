import React from 'react';
declare type Item = {
    id: any;
    name: string;
};
export interface Props {
    maxWidth?: number | string;
    filters?: {
        name: string;
        allItems?: boolean;
        clear?: boolean;
        items?: number[];
    }[];
    onChange?: any;
    value?: Item[];
    items: Item[];
    isLoading?: boolean;
    isFetchable?: boolean;
    placeholder?: string;
    loadingMessage?: string;
    emptyMessage?: string;
    errorMessage?: string;
    errorForm?: boolean;
    isDependent?: boolean;
    dependentMessage?: string;
}
export declare const MultiSelect: React.FC<Props>;
export {};
//# sourceMappingURL=MultiSelect.d.ts.map