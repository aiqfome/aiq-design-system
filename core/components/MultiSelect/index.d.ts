export * from './MultiSelect';
//# sourceMappingURL=index.d.ts.map