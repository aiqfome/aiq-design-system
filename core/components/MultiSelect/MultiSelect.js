"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultiSelect = void 0;
const react_1 = __importDefault(require("react"));
const prop_types_1 = __importDefault(require("prop-types"));
const MultiSelectFetchable_1 = require("./MultiSelectFetchable");
const MultiSelectStatic_1 = require("./MultiSelectStatic");
exports.MultiSelect = (_a) => {
    var { isFetchable = false } = _a, props = __rest(_a, ["isFetchable"]);
    if (isFetchable) {
        return react_1.default.createElement(MultiSelectFetchable_1.MultiSelectFetchable, Object.assign({}, props));
    }
    return react_1.default.createElement(MultiSelectStatic_1.MultiSelectStatic, Object.assign({}, props));
};
exports.MultiSelect.propTypes = {
    isFetchable: prop_types_1.default.bool
};
