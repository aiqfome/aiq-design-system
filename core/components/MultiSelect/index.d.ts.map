{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["../../../src/components/MultiSelect/index.tsx"],"names":[],"mappings":"AAAA,cAAc,eAAe,CAAA"}