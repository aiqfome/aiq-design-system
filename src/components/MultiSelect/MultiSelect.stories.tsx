import React, { ReactElement } from 'react'

import { Flex } from '../Flex'

import { MultiSelect } from './MultiSelect'

export default {
  component: MultiSelect,
  title: 'MultiSelect'
}

const items = [
  { id: 0, name: 'Maringá' },
  { id: 1, name: 'Guarapuava' },
  { id: 2, name: 'São Paulo' },
  { id: 3, name: 'Curitiba' },
  { id: 4, name: 'Cruzeiro do Sul' },
  { id: 5, name: 'Pato Branco' },
  { id: 6, name: 'Prudentópolis' },
  { id: 7, name: 'Campo Mourão' },
  { id: 8, name: 'New York' }
]

const filters = [
  { allItems: true, name: 'todas as cidades' },
  { items: [0, 1, 2, 3, 4], name: 'unidades próprias' },
  { clear: true, name: 'limpar cidades selecionadas' }
]

export const Basic: React.FC = (): ReactElement => {
  function handleChangeMultiSelect(event) {
    console.log(event)
  }

  return (
    <Flex variant='fullCentralized'>
      <MultiSelect
        maxWidth='800px'
        value={[items[0]]}
        onChange={handleChangeMultiSelect}
        filters={filters}
        items={items}
        isLoading={false}
      />
    </Flex>
  )
}
