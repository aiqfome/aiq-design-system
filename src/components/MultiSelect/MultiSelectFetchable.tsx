import React, { useEffect, useRef, useState } from 'react'

import styled from 'styled-components'
import { MdClose } from 'react-icons/md'
import { useCombobox, useMultipleSelection } from 'downshift'

import { Flex } from '../Flex'
import { Box } from '../Box'
import { Text } from '../Text'
import { Divider } from '../Divider'
import { Button } from '../Button'
import { Loading } from '../Loading'
import { InputErrorMessage } from '../InputErrorMessage'

type Item = {
  id: any
  name: string
  select?: string
}

export interface Props {
  maxWidth?: number | string
  filters?: {
    name: string
    allItems?: boolean
    clear?: boolean
    items?: number[]
    onClick?: any
  }[]
  handleSelectedItemChange?: (item: any) => void
  onChange?: any
  value?: Item[]
  items: Item[]
  isLoading?: boolean
  placeholder?: string
  loadingMessage?: string
  emptyMessage?: string
  errorMessage?: string
  errorForm?: boolean
  isDependent?: boolean
  dependentMessage?: string
  disabled?: boolean
}

interface ContainerInputProps {
  errorForm?: boolean
}

const MultiSelectStyled = styled(Box)`
  &:hover {
    cursor: text;
  }

  & input::placeholder {
    color: ${({ theme }) => theme.colors.grey};
    font-size: ${({ theme }) => theme.fontSizes.medium};
  }
`

const ContainerInput = styled(Box)<ContainerInputProps>`
  display: flex;
  flex-direction: row;
  overflow: auto;
  align-items: center;
  padding: 4px 10px;
  border: ${({ errorForm, theme }) =>
    errorForm
      ? `1px solid ${theme.colors.error}`
      : `1px solid ${theme.colors.mediumGrey}`};

  input {
    background: none;
    border: none;
    height: 25px;
    flex: 1;
    font-family: inherit;
  }

  input::placeholder {
    font-family: inherit;
  }

  &::-webkit-scrollbar {
    height: 5px;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 2px rgba(0, 0, 0, 0.3);
  }

  &::-webkit-scrollbar-thumb {
    background-color: darkgrey;
    outline: 1px solid slategrey;
  }
`

interface OverflowProps {
  isOpen?: boolean
}

const Overflow = styled(Flex)<OverflowProps>`
  position: absolute;
  min-width: 100%;
  width: max-content;
  z-index: 99;

  display: ${({ isOpen }) => (isOpen ? 'block' : 'none')};

  ul {
    padding: 0;
    margin: 0;
    list-style: none;

    li {
      padding: 3px 10px;

      &:hover {
        cursor: pointer;
        background: ${({ theme }) => theme.colors.primary};
        color: ${({ theme }) => theme.colors.white};
      }
    }
  }
`

const Itens = styled(Box)`
  max-height: 250px;
  overflow: auto;

  .highlighted {
    background: ${({ theme }) => theme.colors.primary};
    color: ${({ theme }) => theme.colors.white};
  }
`

const SelectedItem = styled(Text)`
  white-space: nowrap;
`

export const MultiSelectFetchable: React.FC<Props> = ({
  items,
  maxWidth,
  filters = [],
  onChange,
  value = [],
  isLoading = false,
  placeholder,
  loadingMessage = 'carregando...',
  emptyMessage = 'item não encontrado ou já adicionado',
  errorForm,
  errorMessage,
  isDependent = false,
  dependentMessage = 'este campo tem alguma dependência',
  disabled,
  handleSelectedItemChange = () => {
    // do nothing.
  },
  ...props
}) => {
  const [inputValue, setInputValue] = useState<string>('')
  const [refBadges, setRefBadges] = useState<HTMLDivElement | null>(null)
  const [itemLimit, setItemLimit] = useState<number | undefined>(undefined)
  const [refContainer, setRefContainer] = useState<HTMLDivElement | null>(null)

  const inputRef = useRef(document.createElement('input'))

  const {
    getSelectedItemProps,
    getDropdownProps,
    selectedItems = []
  } = useMultipleSelection({
    selectedItems: value,
    initialSelectedItems: value,
    onSelectedItemsChange: handleSelectedItemChange
  })

  const getFilteredItems = () =>
    items.filter(item => selectedItems.indexOf(item) < 0)

  const {
    isOpen,
    getMenuProps,
    getInputProps,
    getComboboxProps,
    highlightedIndex,
    getItemProps,
    openMenu
  } = useCombobox({
    inputValue,
    defaultHighlightedIndex: 0,
    selectedItem: null,
    itemToString: item => (item ? '' : ''),
    items: isDependent ? [] : getFilteredItems(),
    stateReducer: (state, actionAndChanges) => {
      const { changes, type } = actionAndChanges
      switch (type) {
        case useCombobox.stateChangeTypes.InputKeyDownEnter:
        case useCombobox.stateChangeTypes.ItemClick:
          return {
            ...changes,
            isOpen: true
          }
      }
      return changes
    },
    onStateChange: ({ inputValue, type, selectedItem }) => {
      switch (type) {
        case useCombobox.stateChangeTypes.InputChange:
          onChange(inputValue)
          setInputValue(inputValue || '')
          break
        case useCombobox.stateChangeTypes.InputKeyDownEnter:
        case useCombobox.stateChangeTypes.ItemClick:
        case useCombobox.stateChangeTypes.InputBlur:
          if (selectedItem) {
            handleSelectedItemChange({
              selectedItems: [...selectedItems, selectedItem]
            })
          }
          onChange('')
          setInputValue('')
          break
        default:
          break
      }
    }
  })

  useEffect(() => {
    const resizeListener = () => setItemLimit(undefined)

    window.addEventListener('resize', resizeListener)
    return () => {
      window.removeEventListener('resize', resizeListener)
    }
  }, [])

  useEffect(() => {
    if (refBadges !== null && refContainer !== null) {
      if (selectedItems.length > 1) {
        const containerSize = refContainer.offsetWidth

        let total = 0
        let limit = -1
        ;(refBadges.childNodes as NodeListOf<HTMLDivElement>).forEach(
          (badge: HTMLDivElement, index) => {
            total = total + badge?.offsetWidth
            if (total >= containerSize * 0.9 - 60 && limit === -1) {
              limit = index
              return false
            }
          }
        )

        if (containerSize * 0.9 - 60 <= total && !itemLimit) {
          if (itemLimit !== limit) setItemLimit(limit)
        }
      }
    }
  }, [
    refBadges,
    itemLimit,
    refContainer,
    selectedItems,
    refContainer?.offsetWidth
  ])

  const clear = () => {
    setItemLimit(undefined)
    handleSelectedItemChange({ selectedItems: [] })
  }

  const filterItems = filter => {
    if (filter.clear) {
      clear()
    }
    if (filter.allItems) {
      handleSelectedItemChange({
        selectedItems: items
      })
    }
    if (filter.items) {
      handleSelectedItemChange({
        selectedItems: items.filter((_, i) => filter?.items?.includes(i))
      })
    }
  }

  return (
    <Flex flexDirection='column' flex={1}>
      <MultiSelectStyled
        position='relative'
        maxWidth={maxWidth}
        data-testid='multiselect-fechable'
        onClick={() => inputRef.current.focus()}
        {...props}
      >
        <ContainerInput
          backgroundColor='white'
          borderRadius={4}
          display='flex'
          flexDirection='row'
          errorForm={errorForm}
          ref={el => {
            getComboboxProps()?.ref(el)
            setRefContainer(el || null)
          }}
        >
          <Flex
            maxWidth='90%'
            overflow='hidden'
            ref={el => setRefBadges(el || null)}
          >
            {selectedItems.slice(0, itemLimit).map((selectedItem, index) => (
              <Flex
                py={2}
                minWidth={selectedItems.length === 1 ? 0 : undefined}
                key={`selected-item-${index}`}
                px={4}
                mr={3}
                display='flex'
                flexDirection='row'
                alignItems='center'
                backgroundColor={disabled ? 'darkGrey' : 'primary'}
                borderRadius='3px'
              >
                <SelectedItem
                  truncate
                  {...getSelectedItemProps({
                    selectedItem,
                    index
                  })}
                  color='white'
                >
                  {selectedItem?.select || selectedItem?.name}
                </SelectedItem>

                <Button
                  disabled={disabled}
                  onClick={e => {
                    e.stopPropagation()
                    handleSelectedItemChange({
                      selectedItems: selectedItems.filter(
                        e => e.id !== selectedItem.id
                      )
                    })
                  }}
                  ml={6}
                >
                  <MdClose color='#fff' />
                </Button>
              </Flex>
            ))}

            {itemLimit !== undefined && selectedItems.length > itemLimit && (
              <Flex
                py={2}
                px={4}
                mr={3}
                display='flex'
                flexDirection='row'
                alignItems='center'
                backgroundColor='primary'
                borderRadius='3px'
              >
                <Text color='white'>
                  {`+${selectedItems.length - itemLimit}`}
                </Text>
              </Flex>
            )}
          </Flex>

          <input
            disabled={disabled}
            type='text'
            placeholder={placeholder}
            style={{ width: '100%', flex: 1, paddingLeft: '5px' }}
            {...getInputProps(
              getDropdownProps({
                ref: inputRef,
                preventKeyAction: isOpen,
                onFocus: () => {
                  if (!isOpen) {
                    openMenu()
                  }
                }
              })
            )}
            autoComplete='disabled'
          />

          {isLoading && <Loading size='small' />}
        </ContainerInput>

        <Overflow
          isOpen={isOpen && !disabled}
          mt={13}
          py={7}
          flexDirection='column'
          backgroundColor='white'
          border='1px solid #dedede'
          {...getMenuProps()}
        >
          {!isDependent && (
            <>
              <ul>
                {filters.map((filter, index) => (
                  <li
                    key={`filter-${index}`}
                    onClick={() => {
                      if (filter.onClick) filter.onClick(filter)
                      filterItems(filter)
                    }}
                  >
                    <Text cursor='pointer'>{filter.name}</Text>
                  </li>
                ))}
              </ul>

              <Divider mx={5} my={4} />
            </>
          )}

          <Itens>
            <ul>
              {isOpen &&
                !isDependent &&
                !isLoading &&
                !disabled &&
                getFilteredItems().map((item, index) => (
                  <li
                    className={highlightedIndex === index ? 'highlighted' : ''}
                    key={`${item}${index}`}
                    onClick={e => {
                      getItemProps({ item, index }).onClick(e)
                      setItemLimit(undefined)
                    }}
                  >
                    {item.name}
                  </li>
                ))}

              {isDependent && <li>{dependentMessage}</li>}

              {isOpen && isLoading && !isDependent && <li>{loadingMessage}</li>}

              {isOpen &&
                !isLoading &&
                !isDependent &&
                getFilteredItems().length === 0 && <li>{emptyMessage}</li>}
            </ul>
          </Itens>
        </Overflow>
      </MultiSelectStyled>

      {errorForm && <InputErrorMessage errorMessage={errorMessage} />}
    </Flex>
  )
}
