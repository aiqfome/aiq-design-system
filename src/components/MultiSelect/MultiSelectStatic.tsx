import React, { useState, useRef } from 'react'
import PropTypes from 'prop-types'

import styled from 'styled-components'
import { MdClose } from 'react-icons/md'
import { useCombobox, useMultipleSelection } from 'downshift'

import { Flex } from '../Flex'
import { Box } from '../Box'
import { Text } from '../Text'
import { Divider } from '../Divider'
import { Button } from '../Button'

type Item = {
  id: any
  name: string
}

export interface Props {
  maxWidth?: number | string
  filters?: {
    name: string
    allItems?: boolean
    clear?: boolean
    items?: number[]
  }[]
  onChange?: any
  value?: Item[]
  items: Item[]
  isLoading?: boolean
  isFetchable?: boolean
}

const MultiSelectStyled = styled(Box)`
  &:hover {
    cursor: text;
  }
`

interface ContainerInputProps {
  onClick: () => void
}

const ContainerInput = styled(Box)<ContainerInputProps>`
  display: flex;
  flex-direction: row;
  overflow: auto;
  align-items: center;
  padding: 4px 10px;
  justify-content: 'space-between';

  input {
    background: none;
    border: none;
    height: 25px;
  }

  &::-webkit-scrollbar {
    height: 5px;
  }

  &::-webkit-scrollbar-track {
    box-shadow: inset 0 0 2px rgba(0, 0, 0, 0.3);
  }

  &::-webkit-scrollbar-thumb {
    background-color: darkgrey;
    outline: 1px solid slategrey;
  }
`

interface OverflowProps {
  isOpen?: boolean
}

const Overflow = styled(Flex)<OverflowProps>`
  position: absolute;
  width: 100%;
  border-top: none;

  display: ${({ isOpen }) => (isOpen ? 'block' : 'none')};

  ul {
    padding: 0;
    margin: 0;
    list-style: none;

    li {
      padding: 3px 10px;

      &:hover {
        cursor: pointer;
        background: ${({ theme }) => theme.colors.primary};
        color: ${({ theme }) => theme.colors.white};
      }
    }
  }
`

const Itens = styled(Box)`
  max-height: 250px;
  overflow: auto;

  .highlighted {
    background: ${({ theme }) => theme.colors.primary};
    color: ${({ theme }) => theme.colors.white};
  }
`

const SelectedItem = styled(Text)`
  white-space: nowrap;
`

export const MultiSelectStatic: React.FC<Props> = ({
  items,
  maxWidth,
  filters = [],
  onChange,
  value = [],
  ...props
}) => {
  const [inputValue, setInputValue] = useState<string | undefined>('')

  const inputRef = useRef(document.createElement('input'))

  const propsMultipleSelection = useMultipleSelection({
    initialSelectedItems: value,
    onSelectedItemsChange: event => {
      onChange && onChange(event)
    }
  })

  const propsCombobox = useCombobox({
    itemToString: item => (item ? '' : ''),
    items: getFilteredItems(),
    stateReducer: (state, actionAndChanges) => {
      const { changes, type } = actionAndChanges
      switch (type) {
        case useCombobox.stateChangeTypes.InputKeyDownEnter:
        case useCombobox.stateChangeTypes.ItemClick:
          return {
            ...changes,
            isOpen: true
          }
      }
      return changes
    },
    onStateChange: ({ inputValue: value, type, selectedItem }) => {
      switch (type) {
        case useCombobox.stateChangeTypes.InputChange:
          setInputValue(value)
          break
        case useCombobox.stateChangeTypes.InputKeyDownEnter:
        case useCombobox.stateChangeTypes.ItemClick:
        case useCombobox.stateChangeTypes.InputBlur:
          if (selectedItem) {
            setInputValue('')
            propsMultipleSelection.addSelectedItem(selectedItem)
          }
          break
      }
    }
  })

  function getFilteredItems() {
    return (
      items &&
      items.filter(
        item =>
          inputValue !== undefined &&
          propsMultipleSelection.selectedItems.indexOf(item) < 0 &&
          item.name.toLowerCase().startsWith(inputValue.toLowerCase())
      )
    )
  }

  function filterItems(filter) {
    if (filter.clear) {
      propsMultipleSelection.selectedItems.forEach(item => {
        propsMultipleSelection.removeSelectedItem(item)
      })
    }
    if (filter.allItems) {
      items.forEach(item => {
        propsMultipleSelection.addSelectedItem(item)
      })
    }
    if (filter.items) {
      filter.items.forEach((_, index) => {
        propsMultipleSelection.addSelectedItem(items[index])
      })
    }
  }

  return (
    <MultiSelectStyled position='relative' maxWidth={maxWidth} {...props}>
      <ContainerInput
        backgroundColor='white'
        borderRadius={4}
        display='flex'
        flexDirection='row'
        onClick={() => {
          inputRef.current.focus()
        }}
        border='1px solid #dedede'
        refBox={propsCombobox.getComboboxProps().ref}
      >
        {propsMultipleSelection.selectedItems.map((selectedItem, index) => (
          <Flex
            py={2}
            key={`selected-item-${index}`}
            px={4}
            mr={3}
            display='flex'
            flexDirection='row'
            alignItems='center'
            backgroundColor='primary'
            borderRadius='3px'
          >
            <SelectedItem
              {...propsMultipleSelection.getSelectedItemProps({
                selectedItem,
                index
              })}
              color='white'
            >
              {selectedItem.name}
            </SelectedItem>

            <Button
              onClick={e => {
                e.stopPropagation()
                propsMultipleSelection.removeSelectedItem(selectedItem)
              }}
              ml={6}
            >
              <MdClose color='#fff' />
            </Button>
          </Flex>
        ))}

        <input
          type='text'
          {...propsCombobox.getInputProps(
            propsMultipleSelection.getDropdownProps({
              ref: inputRef,
              preventKeyAction: propsCombobox.isOpen,
              onFocus: () => {
                if (!propsCombobox.isOpen) {
                  propsCombobox.openMenu()
                }
              }
            })
          )}
        />
      </ContainerInput>

      <Overflow
        isOpen={propsCombobox.isOpen}
        mt={13}
        py={7}
        flexDirection='column'
        backgroundColor='white'
        border='1px solid #dedede'
      >
        <ul>
          {filters.map((filter, index) => (
            <li key={`filter-${index}`} onClick={() => filterItems(filter)}>
              <Text cursor='pointer'>{filter.name}</Text>
            </li>
          ))}
        </ul>

        <Divider mx={5} my={4} />

        <Itens>
          <ul {...propsCombobox.getMenuProps()}>
            {propsCombobox.isOpen &&
              getFilteredItems().map((item, index) => (
                <li
                  className={
                    propsCombobox.highlightedIndex === index
                      ? 'highlighted'
                      : ''
                  }
                  key={`${item}${index}`}
                  {...propsCombobox.getItemProps({ item, index })}
                >
                  {item.name}
                </li>
              ))}
          </ul>
        </Itens>
      </Overflow>
    </MultiSelectStyled>
  )
}

MultiSelectStatic.propTypes = {
  items: PropTypes.array.isRequired,
  maxWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  filters: PropTypes.array,
  onChange: PropTypes.func,
  value: PropTypes.array,
  isLoading: PropTypes.bool
}
