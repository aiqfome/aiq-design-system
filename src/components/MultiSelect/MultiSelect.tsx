import React from 'react'
import PropTypes from 'prop-types'

import { MultiSelectFetchable } from './MultiSelectFetchable'
import { MultiSelectStatic } from './MultiSelectStatic'

type Item = {
  id: any
  name: string
}

export interface Props {
  maxWidth?: number | string
  filters?: {
    name: string
    allItems?: boolean
    clear?: boolean
    items?: number[]
  }[]
  onChange?: any
  value?: Item[]
  items: Item[]
  isLoading?: boolean
  isFetchable?: boolean
  placeholder?: string
}

export const MultiSelect: React.FC<Props> = ({
  isFetchable = false,
  ...props
}) => {
  if (isFetchable) {
    return <MultiSelectFetchable {...props} />
  }

  return <MultiSelectStatic {...props} />
}

MultiSelect.propTypes = {
  isFetchable: PropTypes.bool
}
