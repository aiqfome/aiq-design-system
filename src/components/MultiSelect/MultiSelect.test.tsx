import React from 'react'
import { fireEvent } from '@testing-library/react'

import { MultiSelect } from '../MultiSelect'
import { render } from '../utils/test/render'
import { IoIosArrowDown } from 'react-icons/io'

const greenColor = '#6EC531'

const items = [
  { id: 0, name: 'Maringá (ID 9457)', color: greenColor },
  { id: 1, name: 'Guarapuava (ID 9456)' },
  { id: 2, name: 'São Paulo' },
  { id: 3, name: 'Curitiba' },
  { id: 4, name: 'Cruzeiro do Sul' },
  { id: 5, name: 'Pato Branco' },
  { id: 6, name: 'Prudentópolis' },
  { id: 7, name: 'Campo Mourão' },
  { id: 8, name: 'New York' }
]

const filters = [
  { allItems: true, name: 'todas as cidades' },
  { items: [0, 1, 2, 3, 4], name: 'unidades próprias' },
  { clear: true, name: 'limpar cidades selecionadas' }
]

describe('MultiSelect', () => {
  it('should have to render without crashing', () => {
    const component = render(<MultiSelect items={items} />)
    const select = component.getByTestId('multiselect-static')

    expect(component).toBeTruthy()
    expect(select).toBeTruthy()
  })

  it('should have to render without crashing with isFetchable prop', () => {
    const component = render(<MultiSelect isFetchable items={items} />)
    const select = component.getByTestId('multiselect-fechable')

    expect(component).toBeTruthy()
    expect(select).toBeTruthy()
  })

  it('should render placeholder when have placeholder prop', () => {
    const content = 'My placeholder'
    const { getByTestId } = render(
      <MultiSelect items={items} placeholder={content} />
    )

    const input = getByTestId('multiselect-static').querySelector('input')

    expect(input).toHaveAttribute('placeholder', 'My placeholder')
  })

  it('should have svg when have loading prop', () => {
    const { getByTestId } = render(
      <MultiSelect isFetchable items={items} isLoading />
    )
    const select = getByTestId('multiselect-fechable')
    const svg = select.querySelector('svg')

    expect(svg).toBeInTheDocument()
  })

  it('should have the same amount items of the array', () => {
    const { getByTestId, getAllByTestId } = render(
      <MultiSelect items={items} />
    )
    const select = getByTestId('select-input')

    fireEvent.click(select)
    expect(getAllByTestId('select-item').length).toBe(items.length)
  })

  it('should open dropdown when click him', () => {
    const { getByTestId, container } = render(<MultiSelect items={items} />)

    const select = getByTestId('select-input')
    fireEvent.click(select)

    const list = container.querySelectorAll('li')
    expect(list[0]).toBeVisible()
  })

  it('should keep open dropdown when click a item', () => {
    const handleOnClick = jest.fn()
    const { getByTestId, container } = render(
      <MultiSelect items={items} onChange={handleOnClick} />
    )

    const select = getByTestId('select-input')
    fireEvent.click(select)
    const list = container.querySelectorAll('li')

    expect(list[0]).toBeVisible()
    fireEvent.click(list[0])
  })

  it('should call onChange when click on item in dropdown', () => {
    const handleOnClick = jest.fn()

    const { getByTestId, container } = render(
      <MultiSelect items={items} onChange={handleOnClick} />
    )

    const select = getByTestId('select-input')

    fireEvent.click(select)
    const list = container.querySelectorAll('li')
    fireEvent.click(list[0])

    expect(handleOnClick).toBeCalled()
  })

  it('should have the same amount filters of the array', () => {
    const { getAllByTestId } = render(
      <MultiSelect items={items} filters={filters} />
    )

    expect(getAllByTestId('select-filter').length).toBe(filters.length)
  })

  it('should have the same amount items of the array after the search', () => {
    const { getByTestId, container } = render(<MultiSelect items={items} />)

    fireEvent.change(getByTestId('select-input'), {
      target: { value: 'Maringá' }
    })

    const list = container.querySelectorAll('li')
    expect(list.length).toBe(1)
  })

  it('should be disabled when prop disabled is added', () => {
    const { getByTestId } = render(
      <MultiSelect items={items} disabled={true} />
    )

    const select = getByTestId('select-input')

    expect(select).toBeDisabled()
  })

  it('should have custom color when have color prop', () => {
    const { getByTestId } = render(
      <MultiSelect items={items} value={[items[0]]} />
    )

    const BadgeItem = getByTestId('select-selected-item')
    expect(BadgeItem).toHaveStyle({ backgroundColor: greenColor })
  })

  it('should show suffix when prop is provided', () => {
    const { container } = render(
      <MultiSelect
        items={items}
        value={[items[0]]}
        suffix={<IoIosArrowDown />}
      />
    )

    const suffix = container.querySelector('svg')

    expect(suffix).toBeInTheDocument()
  })

  it('should show limit message when the selected items limit is reached', () => {
    const { container } = render(
      <MultiSelect
        items={items}
        value={[items[0], items[1]]}
        suffix={<IoIosArrowDown />}
        selectedItemsLimit={2}
      />
    )

    const list = container.querySelectorAll('li')
    const firstItemText = list[0].textContent

    expect(list.length).toBe(1)
    expect(firstItemText).toContain('quantidade máxima atingida')
  })

  it('should show correct items result when search input is only numbers', () => {
    const { getByTestId, container } = render(<MultiSelect items={items} />)

    fireEvent.change(getByTestId('select-input'), {
      target: { value: '94' }
    })

    const list = container.querySelectorAll('li')
    expect(list.length).toBe(2)
  })
})
