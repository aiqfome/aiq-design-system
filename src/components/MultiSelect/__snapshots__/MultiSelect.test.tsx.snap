// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must match with the previous snapshot snapshot renders 1`] = `
<div
  className="sc-AxjAm gqOzdO"
>
  <div
    className="sc-AxirZ iBBnaA sc-fzplWN LmPil"
  >
    <div
      className="sc-AxirZ hYdxyZ sc-fznyAO bqyZnM"
      display="flex"
      onClick={[Function]}
    >
      <div
        className="sc-AxjAm khQdjW"
        display="flex"
      >
        <span
          className="sc-AxiKw sc-fzokOt cVcXDe"
          color="white"
          onClick={[Function]}
          onKeyDown={[Function]}
          tabIndex={-1}
        >
          Maringá
        </span>
        <button
          className="sc-AxgMl hzzrVt"
          onClick={[Function]}
        >
          <span
            className="sc-AxiKw gjPZme"
            cursor="pointer"
            fontSize="medium"
          >
            <svg
              color="#fff"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": "#fff",
                }
              }
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
          </span>
        </button>
      </div>
      <input
        aria-autocomplete="list"
        aria-controls="downshift-1-menu"
        aria-labelledby="downshift-1-label"
        autoComplete="off"
        id="downshift-0-input"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        type="text"
        value=""
      />
    </div>
    <div
      className="sc-AxjAm sc-fznKkj eJjBBM"
    >
      <ul>
        <li
          onClick={[Function]}
        >
          <span
            className="sc-AxiKw heDoSu"
            cursor="pointer"
          >
            todas as cidades
          </span>
        </li>
        <li
          onClick={[Function]}
        >
          <span
            className="sc-AxiKw heDoSu"
            cursor="pointer"
          >
            unidades próprias
          </span>
        </li>
        <li
          onClick={[Function]}
        >
          <span
            className="sc-AxiKw heDoSu"
            cursor="pointer"
          >
            limpar cidades selecionadas
          </span>
        </li>
      </ul>
      <hr
        className="sc-AxhCb kYcVnP"
        height="1px"
      />
      <div
        className="sc-AxirZ ksioAa sc-fznZeY kBIupy"
      >
        <ul
          aria-labelledby="downshift-1-label"
          id="downshift-1-menu"
          onMouseLeave={[Function]}
          role="listbox"
        />
      </div>
    </div>
  </div>
</div>
`;
